test/Common/DurableHttpTests.cs

```diff
@@ -310,6 +311,86 @@ public async Task DurableHttpAsync_SynchronousAPI_Returns200(string storageProvi
+                    httpMethod: HttpMethod.Get,
+                    headers: headers,
+                    timeout: TimeSpan.FromMilliseconds(5000));
+
+                string functionName = nameof(TestOrchestrations.CallHttpAsyncOrchestrator);
+                var client = await host.StartOrchestratorAsync(functionName, testRequest, this.output);
+                var status = await client.WaitForCompletionAsync(this.output, timeout: TimeSpan.FromSeconds(400));
+
+                var output = status?.Output;
+                DurableHttpResponse response = output.ToObject<DurableHttpResponse>();
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+                await host.StopAsync();
+            }
+        }
+
+        /// <summary>
+        /// End-to-end test which checks if the CallHttpAsync Orchestrator fails  when the
+        /// HTTP request times out and the CallHttpAsync API throws a TimeoutException.
+        /// </summary>
+        [Theory]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        [MemberData(nameof(TestDataGenerator.GetFullFeaturedStorageProviderOptions), MemberType = typeof(TestDataGenerator))]
+        public async Task DurableHttpAsync_Synchronous_TimeoutException(string storageProvider)
+        {
+            HttpResponseMessage testHttpResponseMessage = CreateTestHttpResponseMessage(HttpStatusCode.OK);
+            HttpMessageHandler httpMessageHandler = MockSynchronousHttpMessageHandlerWithTimeoutException();
+
+            using (ITestHost host = TestHelpers.GetJobHost(
+                this.loggerProvider,
+                nameof(this.DurableHttpAsync_Synchronous_TimeoutException),
+                enableExtendedSessions: false,
+                storageProviderType: storageProvider,
+                durableHttpMessageHandler: new DurableHttpMessageHandlerFactory(httpMessageHandler)))
+            {
+                await host.StartAsync();
+
+                Dictionary<string, string> headers = new Dictionary<string, string>();
+                headers.Add("Accept", "application/json");
+                TestDurableHttpRequest testRequest = new TestDurableHttpRequest(
+                    httpMethod: HttpMethod.Get,
+                    headers: headers,
+                    timeout: TimeSpan.FromMilliseconds(5000));
+
+                string functionName = nameof(TestOrchestrations.CallHttpAsyncOrchestrator);
+                var client = await host.StartOrchestratorAsync(functionName, testRequest, this.output);
+                var status = await client.WaitForCompletionAsync(this.output, timeout: TimeSpan.FromSeconds(400));
+
+                var output = status?.Output;
+                Assert.Contains("System.TimeoutException: The operation was canceled.Reached user specified timeout", output.ToString());
+                Assert.Equal(OrchestrationRuntimeStatus.Failed, status?.RuntimeStatus);
```

Comment:
Right now, the `CallHttpAsync` API throws a `FunctionFailedException` when it reaches a timeout, so that's why these are the Assert conditions that the test is checking. The test should be checking for a `TimeoutException` using `Assert.ThrowsAsync()`. Do you have any recommendations on how to propagate the `TimeoutException`? I went through the stack trace and it looks like it fails to deserialize the exception in TaskActivityDispatcher.cs, which I can't debug.

This is an example of the exception that's thrown when the function reaches the timeout:
```
Function 'Function1 (Orchestrator)' failed with an error. Reason: Microsoft.Azure.WebJobs.Extensions.DurableTask.FunctionFailedException: The activity function 'BuiltIn::HttpActivity' failed: "Failed to deserialize exception from TaskActivity: Unhandled exception while executing task: System.TimeoutException: A task was canceled.Reached user specified timeout: 00:00:10

```

Reply:
So the issue is for some reason TimeoutException is not being serialized properly... We could throw a different exception instead or handle serializing this time of exception more gracefully.

Reply:
Let's try to figure out why it's not being serialized properly. If it's a problem for us then it's likely also a problem for our customers when they have activity functions that throw this exception.